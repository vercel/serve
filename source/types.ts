// source/types.ts
// Type definitions for the CLI.

// An error thrown by any native Node modules.
export declare interface NodeError extends Error {
  code: string;
}

// A path to a file/remote resource.
export declare type Path = string;
// The port to bind the server on.
export declare type Port = number;
// The name of the host.
export declare type Host = string;
// The address of the server.
export declare interface ServerAddress {
  local?: string;
  network?: string;
  previous?: number;
}

// The endpoint the server should listen on.
export declare type ListenEndpoint =
  | number
  | `tcp://${Host}:${Port}`
  | `unix:${Path}`
  | `pipe:\\\\.\\pipe\\${Host}`;

// The parsed endpoints.
export declare interface ParsedEndpoint {
  port?: Port;
  host?: Host;
}

// An entry for URL rewrites.
export declare interface Rewrite {
  source: string;
  destination: string;
}

// An entry for redirecting a URL.
export declare type Redirect = Rewrite & {
  type: number;
};

// An entry to send headers for.
export declare interface Header {
  source: string;
  headers: {
    key: string;
    value: string;
  }[];
}

// The configuration for the CLI.
export declare interface Configuration {
  public: Path;
  cleanUrls: boolean | Path[];
  rewrites: Rewrite[];
  redirects: Redirect[];
  headers: Header[];
  directoryListing: boolean | Path[];
  unlisted: Path[];
  trailingSlash: boolean;
  renderSingle: boolean;
  symlinks: boolean;
  etag: boolean;
}

// The options you can pass to the CLI.
export declare interface Options {
  '--help': boolean;
  '--version': boolean;
  '--listen': ParsedEndpoint[];
  '--single': boolean;
  '--debug': boolean;
  '--config': Path;
  '--no-request-logging': boolean;
  '--no-clipboard': boolean;
  '--no-compression': boolean;
  '--no-etag': boolean;
  '--symlinks': boolean;
  '--cors': boolean;
  '--no-port-switching': boolean;
  '--ssl-cert': Path;
  '--ssl-key': Path;
  '--ssl-pass': string;
}

// The arguments passed to the CLI (the options + the positional arguments)
export declare type Arguments = Partial<Options> & {
  _: string[];
};
