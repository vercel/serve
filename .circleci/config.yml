version: 2
jobs:
  build:
    docker:
      - image: circleci/node:latest

    working_directory: ~/repo

    steps:
      - checkout

      - restore_cache:
          keys:
          - v1-dependencies-{{ checksum "package.json" }}
          # fallback to using the latest cache if no exact match is found
          - v1-dependencies-

      - run: yarn

      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-{{ checksum "package.json" }}

      - run: yarn test

