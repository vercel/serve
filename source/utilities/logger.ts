// source/utilities/logger.ts
// A simple colorized console logger.

/* eslint no-console: 0 */

import chalk from 'chalk';

const http = (...message: string[]) =>
  console.info(chalk.bgBlue.bold(' HTTP '), ...message);
const info = (...message: string[]) =>
  console.info(chalk.bgMagenta.bold(' INFO '), ...message);
const warn = (...message: string[]) =>
  console.error(chalk.bgYellow.bold(' WARN '), ...message);
const error = (...message: string[]) =>
  console.error(chalk.bgRed.bold(' ERROR '), ...message);
const log = console.log;

export const logger = { http, info, warn, error, log };
