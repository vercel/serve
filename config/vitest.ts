// config/vitest.ts
// The vitest configuration file.

import { env } from 'node:process';
import { defineConfig } from 'vitest/config';

// Make sure the output of the CLI is in color, so that it matches the
// snapshots.
env.FORCE_COLOR = 2;

export default defineConfig({
  test: {
    // Collect coverage using C8.
    coverage: { enabled: true },
  },
});
