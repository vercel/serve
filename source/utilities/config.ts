// source/utilities/config.ts
// Parse and return the configuration for the CLI.

import {
  resolve as resolvePath,
  relative as resolveRelativePath,
} from 'node:path';
import { readFile } from 'node:fs/promises';
import Ajv from 'ajv';
// @ts-expect-error No type definitions.
import schema from '@zeit/schemas/deployment/config-static.js';
import { resolve } from './promise.js';
import { logger } from './logger.js';
import type { ErrorObject } from 'ajv';
import type { Configuration, Options, NodeError } from '../types.js';

/**
 * Parses and returns a configuration object from the designated locations.
 *
 * @param presentDirectory - The current working directory.
 * @param directoryToServe - The directory to serve.
 * @param args - The arguments passed to the CLI.
 *
 * @returns The parsed configuration.
 */
export const loadConfiguration = async (
  presentDirectory: string,
  directoryToServe: string,
  args: Partial<Options>,
): Promise<Partial<Configuration>> => {
  const files = ['serve.json', 'now.json', 'package.json'];
  if (args['--config']) files.unshift(args['--config']);

  const config: Partial<Configuration> = {};
  for (const file of files) {
    // Resolve the path to the configuration file relative to the directory
    // with the content in it.
    const location = resolvePath(directoryToServe, file);

    // Disabling the lint rule as we don't want to read all the files at once;
    // if we can retrieve the configuration from the first file itself, we
    // shouldn't waste time and resources fetching the other files too.
    // eslint-disable-next-line no-await-in-loop
    const [error, rawContents] = await resolve<string, NodeError>(
      readFile(location, 'utf8'),
    );
    if (error) {
      if (error.code === 'ENOENT' && file !== args['--config']) continue;
      else
        throw new Error(
          `Could not read configuration from file ${location}: ${error.message}`,
        );
    }

    // The configuration can come from three files in different forms:
    // - now.json: `/now/static`
    // - package.json: `/static`
    // - serve.json: `/`
    type ServeConfigurationFile = Partial<Configuration>;
    interface ManifestConfigurationFile {
      static?: Partial<Configuration>;
    }
    interface NowConfigurationFile {
      now: { static?: Partial<Configuration> };
    }
    type ParseableConfiguration =
      | ServeConfigurationFile
      | ManifestConfigurationFile
      | NowConfigurationFile
      | undefined;

    // Parse the JSON in the file. If the parsed JSON is not an object, or the
    // file does not contain valid JSON, throw an error.
    let parsedJson: ParseableConfiguration;
    try {
      parsedJson = JSON.parse(rawContents) as ParseableConfiguration;
      if (typeof parsedJson !== 'object')
        throw new Error('configuration is not an object');
    } catch (parserError: unknown) {
      throw new Error(
        `Could not parse ${location} as JSON: ${
          (parserError as Error).message
        }`,
      );
    }

    // Check if any of these files have a serve specific section.
    if (file === 'now.json') {
      parsedJson = parsedJson as NowConfigurationFile;
      parsedJson = parsedJson.now.static;
    } else if (file === 'package.json') {
      parsedJson = parsedJson as ManifestConfigurationFile;
      parsedJson = parsedJson.static;
    }
    if (!parsedJson) continue;

    // Once we have found a valid configuration, assign it and stop looking
    // through more configuration files.
    Object.assign(config, parsedJson);

    // Warn the user about using deprecated configuration files.
    if (file === 'now.json' || file === 'package.json')
      logger.warn(
        'The config files `now.json` and `package.json` are deprecated. Please use `serve.json`.',
      );

    break;
  }

  // Make sure the directory with the content is relative to the directoryToServe path
  // provided by the user.
  if (directoryToServe) {
    const staticDirectory = config.public;
    config.public = resolveRelativePath(
      presentDirectory,
      staticDirectory
        ? resolvePath(directoryToServe, staticDirectory)
        : directoryToServe,
    );
  }

  // If the configuration isn't empty, validate it against the AJV schema.
  if (Object.keys(config).length !== 0) {
    const ajv = new Ajv({ allowUnionTypes: true });
    const validate = ajv.compile(schema as object);

    if (!validate(config) && validate.errors) {
      const defaultMessage = 'The configuration you provided is invalid:';
      const error = validate.errors[0] as ErrorObject;

      throw new Error(
        `${defaultMessage}\n${error.message ?? ''}\n${JSON.stringify(
          error.params,
        )}`,
      );
    }
  }

  // Configure defaults based on the options the user has passed.
  config.etag = !args['--no-etag'];
  config.symlinks = args['--symlinks'] || config.symlinks;

  return config;
};
