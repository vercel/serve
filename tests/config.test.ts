// tests/config.test.ts
// Tests for the configuration loader.

import { afterEach, describe, test, expect, vi } from 'vitest';

import { loadConfiguration } from '../source/utilities/config.js';
import { logger } from '../source/utilities/logger.js';
import { Options } from '../source/types.js';

// The path to the fixtures for this test file.
const fixtures = 'tests/__fixtures__/config/';
// A helper function to load the configuration for a certain fixture.
const loadConfig = (
  name: 'valid' | 'invalid' | 'non-existent' | 'deprecated',
  args?: Partial<Options> = {},
) => loadConfiguration(process.cwd(), `${fixtures}/${name}`, args);

afterEach(() => {
  vi.restoreAllMocks();
});

describe('utilities/config', () => {
  // Make sure the configuration is parsed correctly when it is in the
  // `serve.json` file.
  test('parse valid config', async () => {
    const configuration = await loadConfig('valid');
    expect(configuration).toMatchSnapshot();
  });

  // Make sure the configuration is parsed correctly when it is a location
  // specified by the `--config` option.
  test('parse valid config at custom location', async () => {
    const configuration = await loadConfig('custom', {
      '--config': 'config.json',
    });
    expect(configuration).toMatchSnapshot();
  });

  // When the configuration in the file is invalid, the function will throw an
  // error.
  test('throw error if config is invalid', async () => {
    loadConfig('invalid').catch((error: Error) => {
      expect(error.message).toMatch(/invalid/);
    });
  });

  // When no configuration file exists, the configuration should be populated
  // with the `etag` and `symlink` options set to their default values, and
  // the `public` option set to the path of the directory.
  test('return default configuration when no source is found', async () => {
    const configuration = await loadConfig('non-existent');
    expect(configuration).toMatchSnapshot();
  });

  // When the configuration source is deprecated, i.e., the configuration lives
  // in `now.json` or `package.json`, a warning should be printed.
  test('warn when configuration comes from a deprecated source', async () => {
    const consoleSpy = vi.spyOn(logger, 'warn');

    const configuration = await loadConfig('deprecated');
    expect(configuration).toMatchSnapshot();

    expect(consoleSpy).toHaveBeenCalledOnce();
    expect(consoleSpy).toHaveBeenLastCalledWith(
      expect.stringContaining('deprecated'),
    );
  });
});
