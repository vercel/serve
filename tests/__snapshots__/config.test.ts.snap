// Vitest Snapshot v1

exports[`utilities/config > parse valid config 1`] = `
{
  "etag": true,
  "public": "tests/__fixtures__/config/valid/app",
  "renderSingle": true,
  "symlinks": undefined,
}
`;

exports[`utilities/config > parse valid config at custom location 1`] = `
{
  "etag": true,
  "public": "tests/__fixtures__/config/custom/app",
  "renderSingle": true,
  "symlinks": undefined,
}
`;

exports[`utilities/config > return default configuration when no source is found 1`] = `
{
  "etag": true,
  "public": "tests/__fixtures__/config/non-existent",
  "symlinks": undefined,
}
`;

exports[`utilities/config > warn when configuration comes from a deprecated source 1`] = `
{
  "etag": true,
  "public": "tests/__fixtures__/config/deprecated/app",
  "symlinks": undefined,
}
`;
